### PR TITLE
channels last callback

### LINKED ISSUE #15175: Support channels_last
## 🚀 Feature

I'd like to try out some `channels_last` training to see if it improves performance (https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html) 

I'm not entirely sure what the best way to do this with lightning is but I also think it should probably be one of those features that you set on the trainer and it just magically works.

### Motivation

Using the `channels_last` memory format can improve performance in some cases

### Pitch

Add a trainer flag that does whatever is needed for `channels_last` so

```
trainer = pl.Trainer( ..., memory_format='channels_last')
```

or something like that and then before anything happens with training/testing you need to convert the module

```
if self.memory_format == 'channels_last':
    lightning_module = lightning_module.to(memory_format=torch.channels_last)
```

and each batch in the train/test/val loops

```
if self.memory_format == 'channels_last':
    batch = batch.to(memory_format=torch.channels_last)
lightning_module(batch)
```

### Alternatives

I have no idea but I assume I could do this without changing lightning although I'm not sure how yet

### Additional context

I am not sure I will be able to PR this one but I'm not opposed to trying

### PR DESCRIPTION
**Pull request by Pedrexus**
## What does this PR do?

This PR fixes issue #15175 by adding a new feature to the `ChannelsLast` callback. This callback changes the module memory format on setup. This is useful when the user wants to use a different memory format and can lead to a considerable speedup in some cases.

Fixes #15175 

This PR does not introduce any breaking changes.

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by Pedrexus** [comment]
Hello, all. I'm having trouble running `make docs`. I'd be glad if someone could help.

**Comment by awaelchli** [comment]
@Pedrexus This should get you started:

```
pip install -r requirements/pytorch/docs.txt
cd docs/source-pytorch
make html 
open ../build/html/index.html
```

Also, I left a comment here: https://github.com/Lightning-AI/lightning/issues/15175#issuecomment-1566310366

**Review by Borda — DISMISSED** [review]

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 
+- Add enable_autolog_hparams argument to Trainer ([#20593](https://github.com/Lightning-AI/pytorch-lightning/pull/20593))
+
+
+- Added `MemoryFormat` callback to easily change the model memory format ([#15175](https://github.com/Lightning-AI/lightning/pull/17680))
+
+
+- Add `toggled_optimizer(optimizer)` method to the LightningModule, which is a context manager version of `toggle_optimize` and `untoggle_optimizer` ([#20771](https://github.com/Lightning-AI/pytorch-lightning/pull/20771))
+
+
```
Comment (Borda) [inline src/lightning/pytorch/CHANGELOG.md]:
```suggestion
- Added `MemoryFormat` callback to easily change the model memory format ([#15175](https://github.com/Lightning-AI/lightning/pull/17680))
```

File: src/lightning/pytorch/callbacks/memory_format.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from typing import Any, Optional
+
+import torch
+
+import lightning.pytorch as pl
+from lightning.pytorch.callbacks import Callback
+from lightning.pytorch.utilities.rank_zero import rank_zero_warn
+
+
+class MemoryFormat(Callback):
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/memory_format.py:31]: @Pedrexus, just thinking if the name is right, since it rather does `ReorderImageDimensions`
